elyra/pipeline/registry.py

```diff
@@ -125,19 +130,47 @@ def add_component(self, request_body):
         catalog_json = self._get_component_catalog_json()
         catalog_json['components'].append(component_json)
 
-        catalog_dir = os.path.join(os.path.dirname(__file__), "resources")
+        catalog_dir = os.path.join(os.path.dirname(__file__), self._resources_dir)
         catalog_file = os.path.join(catalog_dir, f"{self._type}_component_catalog.json")
-        with open(catalog_file, 'a') as f:
+        with open(catalog_file, 'w') as f:
             f.write(json.dumps(catalog_json))
 
-    def parse_component_details(self, component, component_name):
+    def parse_component_details(self, component, component_name=None):
         """Get component name, id, description for palette JSON"""
-        pass
+        raise NotImplementedError
 
     def parse_component_properties(self, component):
         """Get component properties for properties JSON"""
         raise NotImplementedError
 
+    def get_custom_control_id_and_type(self, parameter_type):
+        # This may not be applicable in every case
+        if parameter_type in ["number", "integer"]:
+            return "NumberControl"
+        elif parameter_type in ["bool", "boolean"]:
+            return "BooleanControl"
+        # elif "array" in parameter_type:
+        #     return "StringArrayControl"
+        else:
+            return "StringControl"
+
+    def return_parameter(self, name, control_id, label, description, data):
```

Comment:
`return_xxx` for a method name is atypical because methods return things.  Might the good 'ole `get_parameter` or `compose_parameter` be more indicative.  I think the latter describes what is taking place in the method and, when called within `build_parameter()` kind of implies the final step.

Reply:
`compose` it is!